### ISSUE TITLE
Item#method_missing shouldn't fail silently

### ISSUE THREAD

**Issue by christopherdbull** [karmi/retire#500]
I think having #method_missing on item return the attribute or nil is the wrong thing to do and can lead to some unnecessary and confusing debugging.  Are they any reasons why it doesn't throw a NoMethodError if the attribute isn't found?

**Comment by karmi** [comment]
Yes. It's inline with other `Hash`-like implementations such as _Hashie_, _Hashr_, etc. Compare:

``` ruby
>> require 'hashie'
=> true
>> h = Hashie::Mash.new foo: 'bar'
=> #<Hashie::Mash foo="bar">
>> h.foo
=> "bar"
>> h.moo
=> nil
```

You're welcome to post your _concrete_ problem with this behaviour.

**Comment by christopherdbull** [comment]
That's fine (well, not really in my opinion, but that's ideological, not practical) if these were just simple objects-as-hashes, but they're pretending (in the context of my use) to be ActiveRecord/Mongoid objects, which don't allow method-missing to fail silently.  This leads to obscure bugs, and, given some of the confusion around the way Tire represents these objects in a rails context (including my own when I first started using the gem), I personally think if these objects are going to use the name of their AR counterparts they should have similar error behaviour.

Beyond that I'm also in the Gary Bernhardt camp of not being a big fan of returning nil as it can create insidious errors that are hard to debug.

**Comment by karmi** [comment]
I understand what you're saying, Christopher. But I don't agree. The `Item` instances returned as results are “lookalikes” or “shadows” of underlying models. They behave in a similar way as Hashes do. If you want to have your real model instances returned, use the `:load` option. If you want to have a different behaviour for result set instances, just use the `:wrapper` option.

> This leads to obscure bugs (...)

Again, please post concrete examples. There aren't many which spring to my mind immediately.

>  given some of the confusion around the way Tire represents these objects in a rails context (...)

There shouldn't be any -- `Items` masquerade as your models within Rails; if you don't like the behaviour, please use the `:load` option, or, if you feel like it, write a `:wrapper` customized for your situation.

**Comment by christopherdbull** [comment]
It's up to you really, I don't think we're going to agree on this but we don't have to - I can easily work around it.

:load is an undesirable performance penalty as I'm sure you're aware so I don't really want to have to hit two data stores.  My concrete example was spending a while yesterday tracking down a bug because instead of throwing an error when I was calling a method that didn't exist on Item (masquerading as AR) it was retuning nil and causing the error to occur away from where it was created.  To me this is bad practice, though I know how common it is in ruby-land, it's one of my biggest gripes.  Also, these objects aren't hashes, they're objects, which means IMO they should behave like objects.

**Comment by christianhellsten** [comment]
I agree with te-chris. I would prefer if accessing missing attributes raised an error instead of failing silently. Similar to:

```
attributes.fetch(method_name.to_sym)
```

There's probably a good reason for not raising errors too, and I can use a proxy object as described here:
https://github.com/karmi/tire/wiki/Data-storage-in-Tire-and-wrapping-the-results

**Comment by karmi** [comment]
Bump @karmi.

**Comment by phoet** [comment]
looks like this is also related to #159
